tests/pack/filelimit.c

```diff
@@ -0,0 +1,36 @@
+#include "clar_libgit2.h"
+#include <git2.h>
+
+void test_pack_filelimit__open_repo_with_1025_packfiles(void)
+{
+	git_repository *repo;
+	git_revwalk *walk;
+	git_oid id;
+	int i;
+
+	/*
+	 * This repository contains 1025 packfiles, each with one commit, one tree,
+	 * and two blobs. The first blob (README.md) has the same content in all
+	 * commits, but the second one (file.txt) has a different content in each
+	 * commit.
+	 */
+	cl_git_pass(git_repository_open(&repo, cl_fixture("1025.git")));
```

Comment:
Instead of adding a new test repository for this, how about we do something similar to the following:

```
void test_pack_filelimit__initialize_tiny(void)
{
        cl_git_pass(git_libgit2_opts(GIT_OPT_SET_MWINDOW_FILES, 1));
}

void test_pack_filelimit__initialize_medium(void)
{
        cl_git_pass(git_libgit2_opts(GIT_OPT_SET_MWINDOW_FILES, 100));
}

void test_pack_filelimit__initialize_huge(void)
{
        cl_git_pass(git_libgit2_opts(GIT_OPT_SET_MWINDOW_FILES, 1024));
}

void test_pack_filelimit__cleanup(void)
{
       cl_git_pass(git_libgit2_opts(GIT_OPT_SET_MWINDOW_FILES, MWINDOW_FILES_DEFAULT));
}

void test_pack_filelimit__revwalk(void)
{
        /* your test with a preexisting repo */
}
```

Reply:
the problem with that approach is that if there were to be a regression, since none of the test repositories have enough packfiles to reproduce the issue (the test repos have a maximum of 4 packfiles), the tests would still pass :(

and changing the open file limits to trigger the issue would have other problems because we would need to keep it so low that accidentally changing the order in which files are opened or closing a file a bit late would cause this test to spuriously fail, making the test unnecessarily fragile.

i haven't found many other ways to test this, but here are a few alternatives:

* create a repository with only ~100 packfiles and change the open file limit in the CI (but that would lead to the tests behaving differently on developers' machines)
* same as above (create a repository with ~100 packfiles), but call `setrlimit` programmatically during the initialization of the test.
* generate the packfiles at runtime (although due to https://github.com/libgit2/libgit2/issues/5399 this behavior is currently sort of slow).

wdyt?

Reply:
Okay, I improved this a bit and added part of your suggestion (although had to drop the 1024 file limit since that would cause Linux machines with default limits to fail the tests) and I'm now asserting that the number of open files is the expected one, but we still need a test repo with a few packfiles: none of the pre-existing repos managed to open more than one packfile at a time.

But with this we can prune the testrepo to a pretty low number of packfiles (say, 16). Would that make it more palatable?

Reply:
> the problem with that approach is that if there were to be a regression, since none of the test repositories have enough packfiles to reproduce the issue (the test repos have a maximum of 4 packfiles), the tests would still pass :(

True, 4 maximum isn't that many. It could be detected in case we set limit to 1, though.

> generate the packfiles at runtime (although due to #5399 this behavior is currently sort of slow).

Did you try using the `git_packbuilder` interface to do this? If we settle for a "sane" limit like generating a 100 packfiles I'd imagine it to not cause much of a slowdown.

> But with this we can prune the testrepo to a pretty low number of packfiles (say, 16). Would that make it more palatable?

Yeah, 16 would definitely be fine with me, if using the packbuilder interface doesn't work.

Reply:
> > the problem with that approach is that if there were to be a regression, since none of the test repositories have enough packfiles to reproduce the issue (the test repos have a maximum of 4 packfiles), the tests would still pass :(
> 
> True, 4 maximum isn't that many. It could be detected in case we set limit to 1, though.

It couldn't, since at most all the pre-existing repositories use one packfile per ref.

> 
> > generate the packfiles at runtime (although due to #5399 this behavior is currently sort of slow).
> 
> Did you try using the `git_packbuilder` interface to do this? If we settle for a "sane" limit like generating a 100 packfiles I'd imagine it to not cause much of a slowdown.

Yes, even with `git_packbuilder`, adding packfiles is potentially quadratic since every time a new packfile appears, the cache is flushed and in the worst case every packfile needs to be read before finding the right one.

Now that I'm explicitly checking how many files are still open, just 16 commits does the job, so the quadratic argument above has been mitigated.

> 
> > But with this we can prune the testrepo to a pretty low number of packfiles (say, 16). Would that make it more palatable?
> 
> Yeah, 16 would definitely be fine with me, if using the packbuilder interface doesn't work.